### PR TITLE
Improve support for directed graphs in A*; docs update included

### LINKED ISSUE #30520: AStar bidirectional argument in methods
Hello, i've been wondering if there could be an additional argument(bidirectional) added to methods "are_points_connected" and "disconnect_points" in AStar. Right now it's possible to create an unidirectional connection between points with "connect_points" and other methods should work likewise.

Example:
If you create unidirectional connection between points 0 and 1 you can't easily remove it or check if there is such a connection. 
Removing connection from 0 to 1 also removes connection from 1 to 0.
Checking for connection from 0 to 1 returns true if there's a connection from 1 to 0.

### PR DESCRIPTION
**Pull request by ayuusweetfish**
Closes #30520. A `bidirectional` parameter is added to methods `are_points_connected()` and `disconnect_points()`. The documentation has been updated to match current method signatures and to be more specific. (ref: #8146)

Current implementation takes about 2x memory for storing edges. This can be optimized if concerns arise on memory usage.

~~I have used `ERR_EXPLAINC` to report warnings, but don't see them anywhere. `DEBUG_ENABLED` is defined. Is that normal?~~ Fixed by using `WARN_PRINT`, rebased already.

<details><summary>Here is a test. Should it be added to main tests?</summary>

```gdscript
var a = AStar.new()
a.add_point(1, Vector3(0, 0, 0))
a.add_point(2, Vector3(0, 1, 0))
a.add_point(3, Vector3(1, 1, 0))
a.add_point(4, Vector3(2, 0, 0))

a.connect_points(1, 2, true)
a.connect_points(1, 3, true)
a.connect_points(1, 4, false)
print('------')
print(a.get_point_connections(1))	# [2, 3, 4]
print(a.get_point_connections(2))	# [1]
print(a.get_point_connections(3))	# [1]
print(a.get_point_connections(4))	# []
print(a.are_points_connected(2, 1))	# True
print(a.are_points_connected(4, 1))	# True
print(a.are_points_connected(2, 1, false))	# True
print(a.are_points_connected(4, 1, false))	# False

a.disconnect_points(1, 2)
print('------')
print(a.get_point_connections(1))	# [3, 4]
print(a.get_point_connections(2))	# []
print(a.get_point_connections(3))	# [1]
print(a.get_point_connections(4))	# []

a.disconnect_points(4, 1, false)
print('------')
print(a.get_point_connections(1))	# [3, 4]
print(a.get_point_connections(2))	# []
print(a.get_point_connections(3))	# [1]
print(a.get_point_connections(4))	# []

a.disconnect_points(4, 1, true)
print('------')
print(a.get_point_connections(1))	# [3]
print(a.get_point_connections(2))	# []
print(a.get_point_connections(3))	# [1]
print(a.get_point_connections(4))	# []

a.connect_points(2, 3, false)
print('------')
print(a.get_point_connections(1))	# [3]
print(a.get_point_connections(2))	# [3]
print(a.get_point_connections(3))	# [1]
print(a.get_point_connections(4))	# []

a.connect_points(2, 3, true)
print('------')
print(a.get_point_connections(1))	# [3]
print(a.get_point_connections(2))	# [3]
print(a.get_point_connections(3))	# [1, 2]
print(a.get_point_connections(4))	# []

a.disconnect_points(2, 3, false)
print('------')
print(a.get_point_connections(1))	# [3]
print(a.get_point_connections(2))	# []
print(a.get_point_connections(3))	# [1, 2]
print(a.get_point_connections(4))	# []

a.remove_point(3)
print('------')
print(a.get_point_connections(1))	# []
print(a.get_point_connections(2))	# []
print(a.get_point_connections(4))	# []
```

</details>

Would love to hear opinions ^ ^

### CODE REVIEW

**Comment by bojidar-bg** [comment]
Does `remove_point` still work?
Also, `get_closest_position_in_segment` would be twice slower after the change, maybe make it assert that `s.from < s.to`?

**Comment by kawa-yoiko** [comment]
> Does `remove_point` still work?

`unlinked_neighbours` are kept valid so I think yes. Maybe an extensive test should be created to confirm that and prevent regression.

> Also, `get_closest_position_in_segment` would be twice slower after the change

Thanks for pointing out! Taking performance and memory use into account, now I think it's better to return to the `p_from < p_to` assertion and just add another flag denoting its direction (either way or bidirectional). I will make changes soon.

**Comment by kawa-yoiko** [comment]
Hi! Worked intermittently on this and I think it's good now.

I have added both manual tests and random batch tests, hope they are strong enough 😝

File: core/math/a_star.cpp [hunk 1]
```diff
@@ -193,10 +218,15 @@ PoolVector<int> AStar::get_point_connections(int p_id) {
...
 
-bool AStar::are_points_connected(int p_id, int p_with_id) const {
+bool AStar::are_points_connected(int p_id, int p_with_id, bool bidirectional) const {
 
 	Segment s(p_id, p_with_id);
-	return segments.has(s);
+	const Set<Segment>::Element *element = segments.find(s);
+	int direction = bidirectional ?
+							0 : // Use 0 to accept any existing segment
+							(s.u == p_id ? Segment::FORWARD : Segment::BACKWARD);
```
Comment (bojidar-bg) [inline core/math/a_star.cpp]: Can't `s.u == p_id ? Segment::FORWARD : Segment::BACKWARD` be replaced by `s.direction`?

**Comment by bojidar-bg** [comment]
Apart from those comments, I think it is good (assuming it works).

**Comment by kawa-yoiko** [comment]
Thanks for your careful review! I have pushed changes and re-run tests.

> assuming it works

Currently the list of neighbours are checked, and that should be enough for now since `_solve()` only depends on them. If desirable, I can add a simple test to generate random graphs and test our A* implementation against Floyd–Warshall.

**Comment by kawa-yoiko** [comment]
I have added an intensive stress test against Floyd–Warshall, which has helped to catch a little bug (diff: [original](https://github.com/godotengine/godot/commit/2e9df83c2c215522a2629b66fde6da7f2006fd9a#diff-d1d91d2dab401f2c55f57b97b6154b2eR171) and [current](https://github.com/godotengine/godot/commit/818aa2c5ad2e276bf899dadb8cadf6602e47c6ab#diff-d1d91d2dab401f2c55f57b97b6154b2eR171)). I also manually merged changes in #30708 and fixed a compilation error on CI.

Hope it's ready now 🚀🚀

File: doc/classes/AStar.xml [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 	<brief_description>
-		AStar class representation that uses 3d-vectors as edges.
+		An implementation of A* to find shortest paths among connected points in space.
 	</brief_description>
 	<description>
-		A* (A star) is a computer algorithm that is widely used in pathfinding and graph traversal, the process of plotting an efficiently directed path between multiple points. It enjoys widespread use due to its performance and accuracy. Godot's A* implementation make use of vectors as points.
-		You must add points manually with [method add_point] and create segments manually with [method connect_points]. So you can test if there is a path between two points with the [method are_points_connected] function, get the list of existing ids in the found path with [method get_id_path], or the points list with [method get_point_path].
+		A* (A star) is a computer algorithm that is widely used in pathfinding and graph traversal, the process of plotting short paths among vertices (points), passing through a given set of edges (segments). It enjoys widespread use due to its performance and accuracy. Godot's A* implementation use 3D points as vertices and use the Euclidean distance by default.
```
Comment (StraToN) [inline doc/classes/AStar.xml]: Godot's A* implementation use -> Godot's A* implementation use**s**

**Review by StraToN — CHANGES_REQUESTED** [review]

**Comment by StraToN** [comment]
Just 1 tiny little error in documentation, all good otherwise for docs.

File: doc/classes/AStar.xml (hunk 2 above)
Comment (kawa-yoiko) [reply doc/classes/AStar.xml]: Haha sure 😝

"Godot's A* implementation uses points in three-dimensional space and Euclidean distances by default."

**Comment by profan** [comment]
@kawa-yoiko I just happened upon this PR that would be quite nice to have indeed, there's some conflicts with changes I've made since but the tests you've written especially would be amazing to have around!

Any chance you'll pick it back up?

**Comment by kawa-yoiko** [comment]
@profan Sure! I probably will check this tomorrow. And thanks for the compliment ^ ^

**Comment by kawa-yoiko** [comment]
@profan It ends up more complicated than I had expected because of bugs in `OAHashMap`... I shall try to understand and fix them soon. Meanwhile, in case you are interested, the rebased branch so far (where tests fail) is [here](https://github.com/godotengine/godot/compare/master...kawa-yoiko:astar-directed-1?expand=1).

**Comment by profan** [comment]
@kawa-yoiko Heh, yes this is part of what I was afraid of, that there would be things that are still subtly broken.. (OAHashMap is after all only employed in A* and the Godot CSG stuff and then basically nowhere else) so it has had quite little testing overall..

**Comment by profan** [comment]
@kawa-yoiko I'm gonna take a look at your rebased branch now btw and try to fix it (OAHashMap), hopefully won't take longer than the weekend, with your battery of tests hopefully it can be worked out 👍

**Comment by kawa-yoiko** [comment]
@profan Such a coincidence that I was just looking into it yesterday and had managed to fix the problem! Will push very soon 😊

**Comment by kawa-yoiko** [comment]
The fix → #32230. Let's wait for that first and rebase from there, so that the engine is 'brought from one stable state to another' 😝

**Comment by kawa-yoiko** [comment]
Rebased and all tests are passing now. (Travis CI somehow failed the Android build because clang++ was not found, which was strange 🤔)

**Comment by profan** [comment]
CC @akien-mga

File: core/math/a_star.cpp [hunk 3]
```diff
@@ -227,10 +250,13 @@ PoolVector<int> AStar::get_point_connections(int p_id) {
 	return point_list;
 }
 
-bool AStar::are_points_connected(int p_id, int p_with_id) const {
+bool AStar::are_points_connected(int p_id, int p_with_id, bool bidirectional) const {
```
Comment (akien-mga) [inline core/math/a_star.cpp]: Should be `p_bidirectional`.

File: core/math/a_star.cpp [hunk 4]
```diff
@@ -164,23 +164,23 @@ void AStar::connect_points(int p_id, int p_with_id, bool bidirectional) {
...
 	segments.insert(s);
 }
 
-void AStar::disconnect_points(int p_id, int p_with_id) {
-
-	Segment s(p_id, p_with_id);
-	ERR_FAIL_COND(!segments.has(s));
-
-	segments.erase(s);
+void AStar::disconnect_points(int p_id, int p_with_id, bool bidirectional) {
```
Comment (akien-mga) [inline core/math/a_star.cpp]: Same here, `p_bidirectional`.

**Comment by akien-mga** [comment]
We're passed release freeze now, but since it was well reviewed and includes tests, I guess we can make an exception to still merge this. I left a nitpick about the naming of a parameter, but let's merge as is to get more extensive testing as early as possible.

**Comment by akien-mga** [comment]
Thanks!

**Comment by kawa-yoiko** [comment]
Ohh glad to see this finally settled! Should there be further issues spotted I shall still be ready to help >ᴗ<

As for parameter naming, will the maintainers team fix it or do I open a new simple PR?

**Comment by akien-mga** [comment]
> As for parameter naming, will the maintainers team fix it or do I open a new simple PR?

There are quite a few occurrences where the convention is not well respected, so we'll do a pass at fixing everything while refactoring code for 4.0 I think.

**Pull request merged**